### ISSUE TITLE
Response to Web Api Controller Action that Returns NotFound() Contains Human-Readable 404 Content

### ISSUE THREAD

**Issue by eriksendc** [aspnet/Mvc#8124]
#### Is this a Bug or Feature request?:
I'm not sure how this issue should be categorized (bug vs feature).

#### Steps to reproduce (preferably a link to a GitHub repo with a repro project):

1. In Startup.cs add this in `Configure()`:

        app.UseStatusCodePagesWithReExecute("/error/{0}");

2. In HomeController.cs, define this action:

        [HttpGet("error/{statusCode}")]
        public IActionResult Error(int statusCode)
        {
            var statusFeature = HttpContext.Features.Get<IStatusCodeReExecuteFeature>();

            if (statusCode == 404 || statusCode == 403)
            {
                return View("NotFound");
            }

            return View();
        }

3. Define NotFound.cshtml in /Views/Shared

4. Define MyApiController.cs that has an action that in some cases returns `NotFound()`.

#### Description of the problem:
My project has both API-related controllers and UI-related controllers. We turned on custom 404 handling to have a good user experience for the regular MVC user experience. We've recently decided to do a more "normal" job in reporting errors from the API controller actions with using BadRequest() and NotFound() to have the HTTP error codes reflect what's gone on.

I understand the call to `NotFound()` is returning a status code of 404. That seems to be intercepted by the error handling middleware (hope I'm saying that right). The net effect is that the body of the response to my API calls is being populated with the HTML of the custom 404 page. Since this is an API call I'd like the payload of the response to just be blank / empty.

Is there some way to make the caller of NotFound() in MyApiController not get intercepted by the middleware so that the caller just gets a blank body? If there isn't a way to do this out-of-the-box I think there should be. I think it's a pretty common pattern to co-mingle API controllers and regular MVC UI page controllers in a single project, right?

#### Version of `Microsoft.AspNetCore.Mvc` or `Microsoft.AspNetCore.App` or `Microsoft.AspNetCore.All`:
`Microsoft.AspNetCore.Mvc` verson is 2.0.1

**Comment by eriksendc** [comment]
Thanks @khellang for pointing me in the right direction. This is one solution I tried based on your comment:

        var statusCodePagesFeature = HttpContext.Features.Get<IStatusCodePagesFeature>();
        if (statusCodePagesFeature != null)
        {
            statusCodePagesFeature.Enabled = false;
        }
        return NotFound();

And this worked. The body of the response was empty.

This got me thinking, though, what would happen if I return some kind of payload by passing an object to `NotFound()`? When I `return BadRequest(ModelState)` I see payload, not HTML of an error page. So I removed the feature disabling code I'd introduced and tried passing an object to `NotFound()`:

        return NotFound(new { PetId = new List<string> { "PetId is invalid." } });

and now the body of the response isn't the HTML of my custom 404, but rather this:

        {
            "PetId": [
                "PetId is invalid."
            ]
        }

I used an anonymous type structured that way so that the errors mimic the format used when you `return BadRequest(ModelState)`. This works for me, as I think it's OK and helpful for the payload of the 404 to give context to what's going on.

Bottom line, it looks like the 404 custom error handler doesn't kick in when some kind of payload is returned.

This is "case closed" for me, but I guess if there was some kind of way for the framework to know that the controller is a Web Api controller instead of an MVC controller, then maybe the parameterless version of `NotFound()` could know to disable the status code pages feature? Anyways, I will leave it to Great Microsoft and Community Minds to decide what to do next.

Thanks again, @khellang !

-Brian Eriksen

**Comment by khellang** [comment]
> Bottom line, it looks like the 404 custom error handler doesn't kick in when some kind of payload is returned.

That's correct. It checks `Response.HasStarted` (headers are flushed and body has been written to) and bails out.

> if there was some kind of way for the framework to know that the controller is a Web Api controller instead of an MVC controller

This is being worked on as we speak with the new "global routing" (formerly known as "dispatcher") feature and will be part of 2.2. It basically lets all middleware know about routes and their metadata so they can act accordingly. Today, this information exist just within MVC.

**Comment by mkArtakMSFT** [comment]
Thanks for handling this, @khellang.
Closing this issue as answered.
